src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -508,16 +508,31 @@ def _visit_where(self, node: foast.Call, **kwargs) -> foast.Call:
             )
 
         try:
-            return_type = type_info.promote(left_type, right_type)
-
-            if isinstance(mask_type, ct.FieldType):
-                if isinstance(return_type, ct.ScalarType):
-                    return_dtype = return_type
-                elif isinstance(return_type, ct.FieldType):
-                    return_dtype = return_type.dtype
-                return_type = type_info.promote(
-                    return_type, ct.FieldType(dims=mask_type.dims, dtype=return_dtype)
-                )
+            if isinstance(left_type, ct.TupleType) and isinstance(right_type, ct.TupleType):
```

Comment:
I know this touches existing code, but I would still suggest changing `left` and `right` to `true_branch` and `false_branch` or similar. It is quite a contained change but a big plus for readability. 

For operators it makes sense to use `left` and `right` because there is no semantic difference between the left and right operand but here we can call things by their meaning.